src/tactic/core.lean

```diff
@@ -573,6 +573,61 @@ do g₀ :: _ ← get_goals,
 
 meta def triv' : tactic unit := do c ← mk_const `trivial, exact c reducible
 
+meta def reorder_goals (gs : list (bool × expr)) : new_goals → list expr
+| new_goals.non_dep_first := (gs.reverse.filter $ coe ∘ bnot ∘ prod.fst).map prod.snd ++ (gs.reverse.filter $ coe ∘ prod.fst).map prod.snd
+| new_goals.non_dep_only := (gs.reverse.filter (coe ∘ bnot ∘ prod.fst)).map prod.snd
+| new_goals.all := gs.reverse.map prod.snd
+
+meta def retry_apply_aux : Π (e : expr) (cfg : apply_cfg), list (bool × expr) → tactic (list (name × expr))
+| e cfg gs :=
+focus1 (do {
+     r ← apply_core e cfg,
+     gs' ← get_goals,
+     set_goals (gs' ++ reorder_goals gs cfg.new_goals),
```

Comment:
Doesn't `apply_core` look at `cfg.new_goals` to order the new goals? Why is the call to `reorder_goals` needed, is there a bug there too?

Reply:
No, the reason is that I simulate part of its behaviour by hand. If `apply_core` does the right thing right away, nothing else needs to happen. But if it doesn't, I give `apply_core` a bit more information and try again. That information would be affected by `apply_core`'s policy on goals but because it's provided as an argument, it will not be treated properly. That's why I explicitly reorder goals.